src/controllers/notifications.js

```diff
@@ -71,6 +73,16 @@ const UnacceptableExpiryError = require('../errors/unacceptable-expiry-error')
 exports.post = function * postNotification () {
   const notification = yield requestUtil.validateBody(this, 'Notification')
   log.debug('Got notification: ' + JSON.stringify(notification))
+
+  const verifySignature = this.config.get('notifications.must_verify')
+  if (verifySignature) {
+    const result = model.verifySignature(notification, this.config)
+    if (!result.valid) {
+      log.warn('Signature verification failed: ' + result.error)
+      throw new UnprocessableEntityError('Notification failed signature verification')
```

Comment:
> If that happens, Ledger should keep retrying (with backoff), to be on the cautious side. So that's why Ledger retries on both 4xx and 5xx today.

The ledger retrying on 4XX errors seems to be a bug according to the definitions of 4XX status codes.

> 10.4.1 400 Bad Request

> The 400 (Bad Request) status code indicates that the server cannot or
   will not process the request due to something that is perceived to be
   a client error (e.g., malformed request syntax, invalid request
   message framing, or deceptive request routing).

Source: https://tools.ietf.org/html/rfc7231#section-6.5

> 11.2.  422 Unprocessable Entity

   > The 422 (Unprocessable Entity) status code means the server
   understands the content type of the request entity (hence a
   415(Unsupported Media Type) status code is inappropriate), and the
   syntax of the request entity is correct (thus a 400 (Bad Request)
   status code is inappropriate) but was unable to process the contained
   instructions.  For example, this error condition may occur if an XML
   request body contains well-formed (i.e., syntactically correct), but
   semantically erroneous, XML instructions.

I haven't seen a proxy server or load balancer return a 4XX when the upstream server is at fault, that's usually what 502, 503, and 504 are reserved for. For example, if the Connector tries to `PUT transfer` on the upstream ledger and fails (gets a 5XX), it should return a `502`. I'm not sure if it does that today.